state-sync/state-sync-v2/state-sync-driver/src/utils.rs

```diff
@@ -58,14 +61,36 @@ pub async fn handle_end_of_stream_or_invalid_payload(
     streaming_service_client: &mut StreamingServiceClient,
     data_notification: DataNotification,
 ) -> Result<(), Error> {
+    // Terminate the stream with the appropriate feedback
     let notification_feedback = match data_notification.data_payload {
         DataPayload::EndOfStream => NotificationFeedback::EndOfStream,
-        _ => NotificationFeedback::InvalidPayloadData,
+        _ => NotificationFeedback::PayloadTypeIsIncorrect,
     };
     terminate_stream_with_feedback(
         streaming_service_client,
         data_notification.notification_id,
         notification_feedback,
     )
-    .await
+    .await?;
+
+    // Return an error if the payload was invalid
+    match data_notification.data_payload {
+        DataPayload::EndOfStream => Ok(()),
+        _ => Err(Error::InvalidPayload("Unexpected payload type!".into())),
+    }
```

Comment:
This is kinda weird cause you could send a response, and get an error back, or get an error back sending even though you would have sent an error at the end anyways.

I'm not sure if it could be cleaner though

Reply:
I'm not sure I understand the exact concern here. When you say `response`, what do you mean? 🤔 This is basically a helper function to handle a data notification. If it's end of stream, it's fine. But if it's not, it means the data type is unexpected, hence the payload.